src/storage/heap_file.c

```diff
@@ -19901,47 +19903,171 @@ heap_build_forwarding_recdes (RECDES * recdes_p, INT16 rec_type, OID * forward_o
  * heap_insert_adjust_recdes_header () - adjust record header for insert
  *                                       operation
  *   thread_p(in): thread entry
- *   context(in): operation context
- *   recdes_p(in): record descriptor to adjust
- *   is_mvcc_op(in): specifies type of operation (MVCC/non-MVCC)
+ *   insert_context(in): insert context
+ *   is_mvcc_class(in): true, if MVCC class
  *   returns: error code or NO_ERROR
  *
- * NOTE: For MVCC operation, it will add an insert_id to the header. For
- *       non-MVCC operations, it will clear all flags.
- * NOTE: The function will alter the provided record descriptor data area.
+ * NOTE: For MVCC class, it will add an insert_id to the header. For non-MVCC class, it will clear all flags.
+ *	 The function will alter the provided record descriptor data area.
  */
 static int
-heap_insert_adjust_recdes_header (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * context, RECDES * recdes_p,
-				  bool is_mvcc_op, bool is_mvcc_class)
+heap_insert_adjust_recdes_header (THREAD_ENTRY * thread_p, HEAP_OPERATION_CONTEXT * insert_context, bool is_mvcc_class)
 {
   MVCC_REC_HEADER mvcc_rec_header;
-  int record_size = recdes_p->length;
-  bool insert_from_reorganize = false;
+  int record_size;
+  int repid_and_flag_bits = 0, mvcc_flags = 0;
+  char *recdes_data, *start_p;
+  MVCCID mvcc_id;
+  bool use_optimization = false;
+
+  assert (insert_context != NULL);
+  assert (insert_context->type == HEAP_OPERATION_INSERT);
+  assert (insert_context->recdes_p != NULL);
+
+  record_size = insert_context->recdes_p->length;
+
+  repid_and_flag_bits = OR_GET_MVCC_REPID_AND_FLAG (insert_context->recdes_p->data);
+  mvcc_flags = (repid_and_flag_bits >> OR_MVCC_FLAG_SHIFT_BITS) & OR_MVCC_FLAG_MASK;
+
+#if defined (SERVER_MODE)
+  use_optimization = is_mvcc_class && (insert_context->update_in_place == UPDATE_INPLACE_NONE)
+    && !heap_is_big_length (record_size + OR_MVCCID_SIZE);
+#endif
+
+  if (use_optimization)
+    {
+      /* 
+       * Most common case. Since is UPDATE_INPLACE_NONE, the header does not have DELID and PREV VERSION.
+       * Optimize header adjustment.
+       */
+      assert (!(mvcc_flags & OR_MVCC_FLAG_VALID_DELID));
+      assert (!(mvcc_flags & OR_MVCC_FLAG_VALID_PREV_VERSION));
+      mvcc_id = logtb_get_current_mvccid (thread_p);
+
```

Comment:
I'd like to propose introducing a pointer variable, `existing_data_p` and renaming `recdes_data` as `new_ins_mvccid_pos_p`.